### PR TITLE
A0-952 calculate totals once

### PR DESCRIPTION
**Pull request by kostekIV**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Infrastructure updated accordingly
- [ ] I have made corresponding changes to the documentation
- [ ] New documentation created
- [ ] Bump `spec_version` and `transaction_version` if relevant
- [ ] Bump `aleph-client` version if relevant

### CODE REVIEW

File: pallets/elections/src/impls.rs [hunk 1]
```diff
@@ -173,19 +176,34 @@ impl<T> Pallet<T>
...
-                ErasReserved::<T>::put(reserved_validators);
-            }
-        }
+        Self::if_era_starts_do(current_era + 1, start_index, || {
+            let reserved_validators = ReservedMembers::<T>::get();
+            ErasReserved::<T>::put(reserved_validators)
+        });
+    }
+
+    fn populate_totals_on_next_era_start(start_index: SessionIndex) {
```
Comment (mike1729) [inline pallets/elections/src/impls.rs:200]: Why `next` in name. It should and works for current era, right? `_on_new_era_start`?

**Review by mike1729 — CHANGES_REQUESTED** [review]
One q and one req for naming convention change.

**Review by maciejnems — APPROVED** [review]
LGTM. Just some nits

File: pallets/elections/src/lib.rs [hunk 2]
```diff
@@ -26,12 +28,20 @@ mod mock;
...
 use frame_support::traits::StorageVersion;
+use scale_info::TypeInfo;
+use sp_std::{collections::btree_map::BTreeMap, prelude::Vec};
+
 pub use pallet::*;
 
 const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
 
 pub type BlockCount = u32;
+pub type Total = u32;
```
[the remark this answers is not in the document]
Comment (kostekIV) [reply pallets/elections/src/lib.rs]: Now it is `TotalReward`

[the remark this answers is not in the document]
Comment (kostekIV) [reply pallets/elections/src/lib.rs]: Now it is `TotalReward`

File: pallets/elections/src/impls.rs [hunk 3]
```diff
@@ -89,12 +90,14 @@ impl<T> Pallet<T>
         <T as pallet_session::Config>::ValidatorId: From<T::AccountId>,
         <T as pallet_session::Config>::ValidatorId: Into<T::AccountId>,
 {
-    fn compute_validator_scaled_totals(era: EraIndex) -> BTreeMap<T::AccountId, u32> {
+    fn compute_validator_scaled_totals(era: EraIndex) {
```
[the remark this answers is not in the document]
Comment (kostekIV) [reply pallets/elections/src/impls.rs]: good catch, renamed

File: pallets/elections/src/impls.rs [hunk 4]
```diff
@@ -173,19 +176,34 @@ impl<T> Pallet<T>
...
 
+    fn if_era_starts_do<F: Fn()>(era: EraIndex, start_index: SessionIndex, on_era_start: F) {
+        if let Some(era_start_index) = pallet_staking::ErasStartSessionIndex::<T>::get(era) {
+            if era_start_index == start_index {
+                on_era_start()
+            }
+        }
+    }
+
     fn populate_reserved_on_next_era_start(start_index: SessionIndex) {
```
[the remark this answers is not in the document]
Comment (kostekIV) [reply pallets/elections/src/impls.rs]: in both function I changed the argument name to just `session`, are you ok with that?

[the remark this answers is not in the document]
Comment (kostekIV) [reply pallets/elections/src/impls.rs]: in both function I changed the argument name to just `session`, are you ok with that?

[the remark this answers is not in the document]
Comment (maciejnems) [reply pallets/elections/src/impls.rs]: I think that `next_index` for `populate_reserved_on_next_era_start` and `start_index` for `populate_totals_on_new_era_start` would be more aligned with how `SessionManager` looks, but just just session is fine with me.

**Review by maciejnems — APPROVED** [review]

**Pull request merged**